pubpub/platform
# feat: add metadata titles to most pages (and rewrite some queries to kysely)

- **feat: add some metadata to a bunch of pages**
- **fix: add titles to stages pages and replace all queries with kysely**
- **feat: improve autocache by React.cache-ing the execute functions**
- **feat: add metadata titles for types and public pages**

## Issue(s) Resolved

All tab titles are simply "PubPub Platform", which makes it hard to distinguish them when you have multiple of them open.

## High-level Explanation of PR


This PR uses Next's metadata API to set the `title` tag on most pages.

There are two ways to do this in Next, statically and dynamically.

Statically it looks like this for e.g. the Members page

https://github.com/pubpub/v7/blob/dd6ad117fffd804aa27e1a0c04371d4739eee398/core/app/c/%5BcommunitySlug%5D/members/%5B%5B...add%5D%5D/page.tsx#L19-L21

You can also dynamically set the metadata by creating a  `generateMetadata` function, like here for the `?editingStageId` pages on `stages/manage`

https://github.com/pubpub/v7/blob/2a37cc588413a9c82f5c2d6541c0817429481f57/core/app/c/%5BcommunitySlug%5D/stages/manage/page.tsx#L27-L47

You might think: Thomas, does this not slow down the generation of the page because we need to make this extra db call every time?

Answer: not necessarily! Because on pages with `?editingStageId` we always use `getStage(editingStageId).executeTakeFirst()`, we can use this "for free" in `generateMetadata` due to `React.cache` deduping multiple function calls\*, plus the fact that they are getting cached by `unstable_cache` in general.

\* Note that this now only works because I actually made `autoCache` use `React.cache` on the entire `execute*` calls, and not just on the part that fetches the data. This means also the query compilation and table-find step are deduped.

https://github.com/pubpub/v7/blob/298b7110b12f671430f2a75807e703a533826480/core/lib/server/cache/autoCache.ts#L18-L56  

 

## Test Plan

1. Go to a bunch of pages, see their title change.
2. Confirm that this does not make loading the pages take significantly longer for e.g. `?editingStageId` pages.

## Screenshots (if applicable)

## Notes
